cachi2/core/utils.py

```diff
@@ -80,11 +84,81 @@ def load_json_stream(s: str) -> Iterator:
         yield obj
 
 
+def _fast_copy(src: Path, dest: Path, *, follow_symlinks: bool = True) -> int:
+    """Perform a fast in-kernel copy using os.copy_file_range syscall.
+
+    Copy data from source path to destination path using a a high-performance copy_file_range(2)
+    syscall. The syscall allows file systems to employ further optimizations like reflinks.
+
+    This should work on Linux >= 4.5 only.
+
+    :param src: source path
+    :param dest: destination path
+    :returns: number of bytes copied
+    """
+
+    def _blocksize(fd: int) -> int:
+        """Determine blocksize for fastcopying on Linux.
+
+        Hopefully the whole file will be copied in a single call.
+        The copying itself should be performed in a loop 'till EOF is
+        reached (0 return) so a blocksize smaller or bigger than the actual
+        file size should not make any difference, also in case the file
+        content changes while being copied.
+        """
+        try:
+            blocksize = max(os.fstat(fd).st_size, 2**23)  # min 8 MiB
+        except Exception:
+            # fall back to 128MiB block
+            blocksize = 2**27
+
+        # On 32-bit architectures truncate to 1 GiB to avoid OverflowError
+        if sys.maxsize < 2**32:
+            blocksize = min(blocksize, 2**30)
```

Comment:
How about using constants instead of comments? For example:
```
MIB_8 = 2**23
MIB_128 = 2**27
```

Reply:
I guess I can do that, there's no harm, although I wanted to keep the code as authentic as possible since I've taken it from elsewhere (referenced and credited) and wanted to minimize the differences for practical reasons.

Reply:
I don't know what is the correct way of doing it...
Personally, I wouldn't have a problem with more significant changes compared to the implementation in CPython.

Should it be referenced in the actual function itself, or is the commit message enough?

Reply:
> Should it be referenced in the actual function itself, or is the commit message enough?

That's what commit messages are for, anyone should be capable of running blame in git and that's where they'd look for more information.

Reply:
> Personally, I wouldn't have a problem with more significant changes compared to the implementation in CPython.

The commentaries are harmless and don't require defining constants that are only going to be used here OR think of it this way - if we ever need to change the value, we need to change it at 2 places, first the value itself and then the name of the constant and all places that use it, so creating constants whose name depend on the value isn't IMO the greatest approach.

Reply:
I get that, but on the other hand, by using comments, we would need to change the comment and then the actual math expression. So, I don't see this as a better approach.